app/src/main/java/com/asfoundation/wallet/transactions/PerkBonusAndGamificationService.kt

```diff
@@ -0,0 +1,290 @@
+                    allLevels: Levels, transactions: List<Transaction> ->
+          if (stats.status != GamificationStats.Status.OK && !stats.isActive &&
+              transactions.isNotEmpty())
+            return@Function5
+          val maxBonus = allLevels.list.maxBy { level -> level.bonus }!!.bonus
+          val maxLevel = allLevels.list.maxBy { level -> level.level }!!.level
+          val currentLevel = stats.level
+          val bonusTransactionValue = getAllPerkBonusTransactionValues(transactions)
+          val currLevelStartAmount = allLevels.list[currentLevel].amount
+          if (lastShownLevel < currentLevel && hasPurchaseResultedInLevelUp(transactions,
+                  stats.totalSpend.minus(currLevelStartAmount))) {
+            promotionsRepository.shownLevel(address, currentLevel,
+                GamificationContext.NOTIFICATIONS_LEVEL_UP.toString())
+            buildNotification(createLevelUpNotification(stats, maxBonus,
+                currentLevel == maxLevel, bonusTransactionValue),
+                NOTIFICATION_SERVICE_ID_PERK_AND_LEVEL_UP)
+          } else if (bonusTransactionValue.isNotEmpty()) {
+            buildNotification(createPerkBonusNotification(bonusTransactionValue),
+                NOTIFICATION_SERVICE_ID_PERK_AND_LEVEL_UP)
+          }
+          if (currentLevel in (almostNextLevelLastShown + 1) until maxLevel) {
+            // the current level being shown is the value to be stored regarding if the notification
+            // for almost reaching next level (current level + 1) has been shown or not
+            val almostNextLevelPercent = gamificationMapper
+                .mapAlmostNextLevelUpPercentage(stats.level)
+            val totalAppCoinsAmountThisLevel = stats.nextLevelAmount!!.minus(
+                currLevelStartAmount)
+            val currentAppCoinsAmountThisLevel = stats.totalSpend.minus(
+                currLevelStartAmount)
+            if (totalAppCoinsAmountThisLevel > BigDecimal.ZERO &&
+                currentAppCoinsAmountThisLevel > BigDecimal.ZERO &&
+                (currentAppCoinsAmountThisLevel.toDouble() / totalAppCoinsAmountThisLevel.toDouble()
+                    * 100.0).toInt() > almostNextLevelPercent) {
+              promotionsRepository.shownLevel(address, currentLevel,
+                  GamificationContext.NOTIFICATIONS_ALMOST_NEXT_LEVEL.toString())
+              buildNotification(createAlmostNextLevelNotification(
+                  formatter.formatGamificationValues(totalAppCoinsAmountThisLevel
+                      .minus(currentAppCoinsAmountThisLevel)), maxBonus),
+                  NOTIFICATION_SERVICE_ID_ALMOST_LEVEL_UP)
+            }
+          }
+        }).subscribe({}, { it.printStackTrace() })
+  }
+
+  private fun getNewTransactions(address: String, timesCalled: Int = 0): List<Transaction> {
+    try {
+      val transactions = transactionRepository.fetchNewTransactions(address).blockingGet()
```

Comment:
Since this will be use in a single, maybe you can try and remove the blockingGet and turn this method into a single